pkg/controller/pxc/version.go

```diff
@@ -0,0 +1,142 @@
+package pxc
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	v1 "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"github.com/robfig/cron/v3"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const jobName = "ensure-version"
+const never = "Never"
+const disabled = "Disabled"
+
+func (r *ReconcilePerconaXtraDBCluster) deleteEnsureVersion(id int) {
+	r.crons.crons.Remove(cron.EntryID(id))
+	delete(r.crons.jobs, jobName)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) sheduleEnsurePXCVersion(cr *api.PerconaXtraDBCluster, vs VersionService) error {
+	if cr.Spec.UpdateStrategy != v1.SmartUpdateStatefulSetStrategyType ||
+		cr.Spec.UpgradeOptions.Schedule == "" ||
+		cr.Spec.UpgradeOptions.Apply == never ||
+		cr.Spec.UpgradeOptions.Apply == disabled {
+		return nil
+	}
+
+	shedule, ok := r.crons.jobs[jobName]
+	if ok && (cr.Spec.UpgradeOptions.Schedule == "" ||
+		cr.Spec.UpgradeOptions.Apply == never ||
+		cr.Spec.UpgradeOptions.Apply == disabled) {
+		r.deleteEnsureVersion(shedule.ID)
+		return nil
+	}
+	if ok && shedule.CronShedule == cr.Spec.UpgradeOptions.Schedule {
+		return nil
+	}
+	if ok {
+		log.Info(fmt.Sprintf("remove job %s because of new %s", shedule.CronShedule, cr.Spec.UpgradeOptions.Schedule))
+		r.deleteEnsureVersion(shedule.ID)
+	}
+
+	log.Info(fmt.Sprintf("add new job: %s", cr.Spec.UpgradeOptions.Schedule))
+	id, err := r.crons.crons.AddFunc(cr.Spec.UpgradeOptions.Schedule, func() {
+		localCr := &api.PerconaXtraDBCluster{}
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: cr.Name, Namespace: cr.Namespace}, localCr)
+		if err != nil {
+			log.Error(err, "failed to get CR")
+			return
+		}
+
+		if localCr.Status.Status != v1.AppStateReady {
+			log.Info("cluster is not ready")
+			return
+		}
```

Comment:
maybe this check should be moved into `ensurePXCVersion`, the reason it is also needed for "entermediate" version.

Reply:
If I move this to `ensurePXCVersion` it will not work because cluster is not ready when we deploying it.

Reply:
can we improve it (like following) and move into `ensurePXCVersion`?
```suggestion
		if localCr.Status.PXC.Version != "" && localCr.Status.Status != v1.AppStateReady {
			log.Info("cluster is not ready")
			return
		}
```

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/416/commits/4dfe3971a01cc41001e3f207ed9b88fb19248a71